src/webgpu/util/conversion.ts

```diff
@@ -755,8 +771,69 @@ export function TypeMat(cols: number, rows: number, elementType: ScalarType): Ma
   return ty;
 }
 
-/** Type is a ScalarType, VectorType, or MatrixType. */
-export type Type = ScalarType | VectorType | MatrixType;
+/** StructType describes the type of WGSL Struct. */
+export class StructType {
+  readonly name: string;
+  readonly elementTypes: Type[]; // Element types
+
+  constructor(name: string, types: Type[]) {
+    this.name = name;
+    this.elementTypes = types;
+  }
+
+  /**
+   * @returns a Struct constructed from the values read from the buffer at the
+   * given byte offset
+   */
+  public read(buf: Uint8Array, offset: number): Struct {
+    const elements: Value[] = this.elementTypes.map(e => {
+      const result = e.read(buf, offset);
+      offset += e.size;
+      // vec3s have a padding element
+      if (e instanceof VectorType) {
+        if (e.width === 3) {
+          offset += e.elementType.size;
```

Comment:
I am afraid that a vec3 in structure doesn't take up the padding space, see [the example in WGSL spec](https://gpuweb.github.io/gpuweb/wgsl/#example-fc0bb4df).
For example, in the following structure
``` rust
struct A {                            //            align(16) size(16)
    a: vec3<f32>,                     // offset(0)  align(16) size(12)
    b: f32,                           // offset(12) align(4)  size(4)
}
```
there is not padding space between elements `a` and `b`. Actually, the spec says ["Each subsequent member is placed at the lowest offset that satisfies the member type alignment, and which avoids overlap with the previous member"](https://gpuweb.github.io/gpuweb/wgsl/#example-fc0bb4df:~:text=Each%20subsequent%20member%20is%20placed%20at%20the%20lowest%20offset%20that%20satisfies%20the%20member%20type%20alignment%2C%20and%20which%20avoids%20overlap%20with%20the%20previous%20member.), and the "overlap" here only consider the actual size of previous member.

So, if I get it right, instead of adding offset after a vec3, we should align the offset before reading each element in line 790. I thought the `modf` vec3 testcases passed as expected is because the result structure is
```rust
struct ModfOutput {
    a: vec3<f32>,                                       // offset(0)  align(16)  size(12)
    // -- implicit vec3<f32> alignment padding for b -- // offset(12)            size(4)
    b: vec3<f32>,                                       // offset(16) align(16)  size(12)
}
```
and the alignment padding size is coincidently equal to the "padding element" size.